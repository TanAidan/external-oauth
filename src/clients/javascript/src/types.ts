declare namespace Types {
  export interface ConnectOptions {
    authId?: string
    setupId?: string
    params?: any
  }

  export interface ConnectSuccess {
    authId: string
  }

  export interface ConnectError extends Error {}

  export interface IntegrationOptions {
    authId?: string
    setupId?: string
  }

  export type RequestHeaders = Record<string, string | number | undefined>
  export type RequestMethod = 'GET' | 'DELETE' | 'HEAD' | 'OPTIONS' | 'POST' | 'PUT' | 'PATCH'

  export type RequestQueryString = Record<string, string | number>
  export interface RequestParameters {
    headers?: RequestHeaders
    query?: RequestQueryString
    body?: any
  }
}

export default Types
