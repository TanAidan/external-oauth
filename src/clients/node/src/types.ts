declare namespace Types {
  export interface IntegrationOptions {
    authId?: string
    setupId?: string
  }

  export type RequestHeaders = Record<string, string | number | undefined>
  export type RequestMethod = 'GET' | 'DELETE' | 'HEAD' | 'OPTIONS' | 'POST' | 'PUT' | 'PATCH'

  export type RequestQueryString = Record<string, string | number>
  export interface RequestParameters {
    headers?: RequestHeaders
    query?: RequestQueryString
    body?: any
  }
}

export default Types
