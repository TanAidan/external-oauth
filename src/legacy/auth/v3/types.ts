import { Request } from 'express'
import Integration from '../../functions/integration'
import Knex from 'knex'

export interface IAuthContext {
  clientId: string
  connectParams: any
  buid?: string
  setupId: string
  isCallback?: boolean
  store: Knex
  configuration: SetupRecord
}

interface SetupRecord {
  setup_id: string
  scopes: string[] | undefined
  buid: string
  configuration: any
}

interface IContextSessionData {
  session: {
    context: any
  }
}

export enum EDashboardApiErrorCodes {
  ENVIRONMENT_NOT_FOUND = 'ENVIRONMENT_NOT_FOUND',
  INTEGRATION_NOT_FOUND = 'INTEGRATION_NOT_FOUND'
}

export interface IDashboardApiResponse {
  errors: [
    {
      code: EDashboardApiErrorCodes
      message: string
    }
  ]
  data: {
    viewer: {
      canInvoke: boolean
      hasAccessTo: boolean
      clientId: string
      setupId: string
      buid: string
    }
  }
}

export type TConnectContextRequest = Request &
  IContextSessionData &
  IAuthContext & {
    params: {
      buid: string
    }
    query: {
      clientId: string
      setupId: string
    }
    integration?: Integration
    fullLogs?: boolean
  }

export type TCallbackContextRequest = Request & IContextSessionData & IAuthContext

export enum ELocalRedirectMethod {
  Inline = 'inline',
  Localhost = 'localhost'
}

type TNameValues = Record<string, any>

export type TIntegrationConfig = TNameValues & {
  authType: EAuthType
  config?: {
    scope?: string[]
    state?: any
  }
}

export interface IAuthConfig {
  integrationConfig: TIntegrationConfig
  setupDetails: TNameValues
}

export enum EAuthType {
  NoAuth = 'NO_AUTH',
  OAuth1 = 'OAUTH1',
  OAuth2 = 'OAUTH2'
}

interface IAuthConfigSessionData {
  session: {
    authConfig: IAuthConfig
  }
}

export type TConnectConfigRequest = Request & IAuthContext & IAuthConfig & IAuthConfigSessionData

export type TCallbackConfigRequest = Request & IAuthConfig & IAuthConfigSessionData

export interface IAuthId {
  authId: string
}

export type TAuthIdRequest = Request &
  IAuthContext &
  IAuthId & {
    session: {
      authId: string
      context: {
        setupId: string
      }
    }
  }

export interface IOAuth1Credentials {
  accessToken: string
  tokenSecret: string
  consumerKey: string
  consumerSecret: string
  expiresIn: number
}

export interface IOAuth2Credentials {
  accessToken: string
  refreshToken: string
  idToken: string
  idTokenJwt?: any
  expiresIn: number
}

export interface IWithAuthCredentials {
  credentials: IOAuth1Credentials | IOAuth2Credentials
  tokenResponse?: OAuthTokenResponse
}

export type TAuthenticateRequest = Request & IAuthContext & IAuthConfig & IWithAuthCredentials

export type AuthSuccessRequest = Request &
  IAuthContext &
  IAuthId &
  IWithAuthCredentials & { buid: string; integrationConfig?: TIntegrationConfig } & { store: Knex }

export type TRevokeRequest = Request & { buid: string }

export type TErrorHandlerRequest = Request & Partial<IAuthContext> & Partial<IAuthId>

export interface IFetchAuthDetailsParams {
  buid: string
  scopedUserDataTableName: string
  servicesTableName: string
  integration: Integration
  environmentIdentifier: string
  authId?: string
  setupId?: string
  configuration: any
  setupIdFromRequest?: boolean
  store: any
}

export interface BasicAuthDetails {
  username: string
  password: string
}

export interface ApiKeyAuthDetails {
  apiKey: string
}

export enum OAuth1SignatureMethod {
  HmacSha1 = 'HMAC-SHA1',
  RsaSha1 = 'RSA-SHA1',
  PlainText = 'PLAINTEXT'
}

interface OAuthAuthDetails {
  callbackParams: Record<string, string>
  tokenResponse?: OAuthTokenResponse
  updatedAt: number
}

export interface OAuth1AuthDetails extends OAuthAuthDetails {
  accessToken: string
  expiresIn: number
  tokenSecret: string
  consumerKey: string
  consumerSecret: string
  signatureMethod: 'HMAC-SHA1' | 'PLAINTEXT' | 'RSA-SHA1'
}

export interface OAuth2AuthDetails extends OAuthAuthDetails {
  accessToken: string
  callbackParams: Record<string, string>
  clientId: string
  clientSecret: string
  expiresIn: number
  idToken?: string
  idTokenJwt?: any
  refreshToken?: string
}

export interface OAuthTokenResponse {
  body: any
  headers: Record<string, string>
}

export type AuthDetails = BasicAuthDetails | ApiKeyAuthDetails | OAuth1AuthDetails | OAuth2AuthDetails | undefined
