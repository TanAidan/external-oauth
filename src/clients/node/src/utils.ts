import Types from './types'

/**
 * Helper to build a new URL from different params
 */
export function toURL(origin: string, baseURL: string, endpoint: string, queryString?: Types.RequestQueryString): URL {
  const removeLeadingSlash = (text: string) => {
    return text.replace(/^\//, '')
  }

  const removeTrailingSlash = (text: string) => {
    return text.replace(/\/$/, '')
  }

  const urlParts: string[] = []
  urlParts.push(removeTrailingSlash(origin))
  urlParts.push(removeLeadingSlash(removeTrailingSlash(baseURL)))
  urlParts.push(removeLeadingSlash(endpoint))

  const url = new URL(urlParts.join('/'))

  if (queryString) {
    Object.keys(queryString).forEach(key => url.searchParams.append(key, String(queryString[key])))
  }

  return url
}

/**
 * Helper to remove all undefined keys
 * @param obj {object}
 */
export function cleanHeaders(obj: Record<string, any>) {
  return Object.keys(obj).reduce((acc, key: string) => {
    if (obj[key] !== undefined) {
      acc[key] = obj[key]
    }
    return acc
  }, {} as Record<string, any>)
}
