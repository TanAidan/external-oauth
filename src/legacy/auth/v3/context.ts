import { TConnectContextRequest, TCallbackContextRequest } from './types'
import { NextFunction, Response } from 'express'
import { UserError } from '../../errors'
import { NoAuthInProgress } from './errors'
import { asyncMiddleware } from '../../errorHandler'

const getConnectParams = (params?: any) => {
  if (params === undefined) {
    return {}
  }

  if (typeof params !== 'object' || Array.isArray(params)) {
    throw new InvalidConnectParams()
  }

  for (const [name, value] of Object.entries(params)) {
    if (typeof value !== 'string' || !/^[\w\s.-]*$/.test(value)) {
      throw new InvalidConnectParam(name)
    }
  }

  return params
}

export const connectContext = (req: TConnectContextRequest, res: Response, next: NextFunction) => {
  const { buid, setupId } = req

  const connectParams = getConnectParams(req.query.params)

  req.session.context = {
    connectParams,
    setupId,
    buid
  }

  req.connectParams = connectParams

  next()
}

export const callbackContext = asyncMiddleware(
  async (req: TCallbackContextRequest, _res: Response, next: NextFunction) => {
    if (!req.session.context) {
      throw new NoAuthInProgress()
    }

    const { buid, connectParams } = req.session.context

    req.isCallback = true
    req.connectParams = connectParams
    req.buid = buid
    req.setupId = req.session.authConfig.setupDetails.setup_id

    next()
  }
)

class InvalidConnectParams extends UserError {
  constructor() {
    super(
      // tslint:disable:max-line-length
      `Incorrect format for connect parameters'

Connect parameters must be sent as query parameters of the form \`params[name]=value\` eg. \`params[subdomain]=my-app\`'`,
      // tslint:enable:max-line-length
      400,
      'INVALID_CONNECT_PARAMS'
    )
  }
}

class InvalidConnectParam extends UserError {
  constructor(name: string) {
    super(
      `Incorrect format for connect parameter '${name}'

Connect parameters may contain alphanumeric and space characters, or any of the following symbols '_-.'

Refer to this link for further information: https://docs.bearer.sh/faq/connect-button`,
      400,
      'INVALID_CONNECT_PARAM'
    )
  }
}
