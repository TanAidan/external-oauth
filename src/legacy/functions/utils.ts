// Is the given object empty e.g. {}?
export function isEmpty(obj: Object) {
  return Object.getOwnPropertyNames(obj).length === 0
}
