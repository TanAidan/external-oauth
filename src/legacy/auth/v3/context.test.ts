import { MiddlewareTestHarness, session } from '../../../../tests/utils'
import { TConnectContextRequest, TCallbackContextRequest } from './types'
import { connectContext, callbackContext } from './context'

const buid = 'test-buid'
const clientId = 'test-client-id'
const setupId = 'test-setup-id'
const envId = 'test-env-123456'
const userCorrelationId = 'user-corr-123'
const internalCorrelationId = 'int-corr-456'
const queryStringBase = `clientId=${clientId}&setupId=${setupId}&correlationId=${userCorrelationId}`
const queryString = `${queryStringBase}&params[conParam]=test-con-param`
const connectParams = { conParam: 'test-con-param' }

describe('connectContext', () => {
  const setup = () =>
    new MiddlewareTestHarness<TConnectContextRequest>({
      setupMiddlewares: [session()],
      testMiddleware: connectContext,
      pathParams: ['buid'],
      configureRequest: (req, _res) => {
        req.buid = 'test-buid'
        req.session.context = {} as any
        req.setupId = setupId
      }
    })

  it('sets up the logger from the request', async () => {
    const test = setup()
    await test
      .get(buid)
      .query(queryString)
      .expect(200)
  })

  it('saves the `params` values as the connect params', async () => {
    const test = setup()

    await test
      .get(buid)
      .query(queryString)
      .expect(200)

    expect(test.req.connectParams).toEqual(connectParams)
  })

  it('validates the format of the connect params', async () => {
    const test = setup()

    await test
      .get(buid)
      .query(`${queryStringBase}&params=oops`)
      .expect(400)

    // tslint:disable:max-line-length
    expect(test.err).toMatchInlineSnapshot(`
[InvalidConnectParams: Incorrect format for connect parameters'

Connect parameters must be sent as query parameters of the form \`params[name]=value\` eg. \`params[subdomain]=my-app\`']
`)
    // tslint:enable:max-line-length
  })

  it('validates the format of an individual connect param', async () => {
    const test = setup()

    await test
      .get(buid)
      .query(`${queryString}&params[invalidParam]=http://naughty.example.com`)
      .expect(400)

    expect(test.err).toMatchInlineSnapshot(`
[InvalidConnectParam: Incorrect format for connect parameter 'invalidParam'

Connect parameters may contain alphanumeric and space characters, or any of the following symbols '_-.'

Refer to this link for further information: https://docs.bearer.sh/faq/connect-button]
`)
  })

  describe('when assigning clientId from query param', () => {
    it('sets the clientId using the environment id', async () => {
      const test = setup()

      await test
        .get(buid)
        .query(queryString)
        .expect(200)

      expect(test.req.clientId).toBe(envId)
    })

    it('raises an error if the clientId query parameter is not present', async () => {
      const test = setup()

      await test.get(buid).expect(400)

      expect(test.err).toMatchSnapshot()
    })
  })

  it('stores the context data in the session', async () => {
    const test = setup()

    await test
      .get(buid)
      .query(queryString)
      .expect(200)

    expect(test.req.session.context).toMatchSnapshot()
  })
})

describe('callbackContext', () => {
  const environmentIdentifier = 'test-643'
  const organizationIdentifier = 'test-5463738'
  const setup = (localAuth = false) =>
    new MiddlewareTestHarness<TCallbackContextRequest>({
      configureRequest: req => {
        req.session.context = {
          clientId,
          connectParams,
          setupId,
          buid,
          environmentIdentifier,
          internalCorrelationId,
          organizationIdentifier,
          userCorrelationId,
          localAuth
        }
      },
      setupMiddlewares: [session()],
      testMiddleware: callbackContext
    })

  it('loads the context from the session', async () => {
    const test = setup()

    await test.get().expect(200)

    expect(test.req).toMatchObject({
      setupId,
      buid,
      connectParams,
      localAuth: false,
      clientId: environmentIdentifier
    })
  })

  it('raises a NoAuthInProgress error if there is no session data', async () => {
    const test = new MiddlewareTestHarness({
      setupMiddlewares: [session()],
      testMiddleware: callbackContext
    })

    await test.get().expect(422)

    expect(test.err).toMatchSnapshot()
  })
})
