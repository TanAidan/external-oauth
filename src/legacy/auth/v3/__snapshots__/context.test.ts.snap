// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`callbackContext raises a NoAuthInProgress error if there is no session data 1`] = `[NoAuthInProgress: The callback was invoked whilst no authentication was in progress. Either too much time has elapsed between starting authentication and the callback being invoked, or the callback was invoked without first calling the integration's authentication endpoint. Please try again]`;

exports[`connectContext stores the context data in the session 1`] = `
Object {
  "aliasBuid": "test-alias",
  "buid": "test-buid",
  "clientId": "test-client-id",
  "connectParams": Object {
    "conParam": "test-con-param",
  },
  "environmentIdentifier": "test-env-123456",
  "localAuth": false,
  "organizationIdentifier": "test-org-12345",
  "setupId": "test-setup-id",
}
`;

exports[`connectContext when assigning clientId from query param raises an error if the clientId query parameter is not present 1`] = `[MissingParameter: 'clientId' must be set]`;
